functorch/_src/make_functional.py

```diff
@@ -241,79 +192,175 @@ def fun(weights, buffers, data):
     return weights, buffers, fun, weight_descriptors, buf_descriptors
 
 
+def make_split_names(lst):
+    return [name.split('.') for name in lst]
+
+
 class FunctionalModuleWithBuffers(nn.Module):
     """
     This is the callable object returned by :func:`make_functional_with_buffers`.
     """
 
-    def __init__(self, stateless_model, param_names, buffer_names,
-                 param_names_map, buffer_names_map):
+    def __init__(
+        self,
+        stateless_model,
+        param_module_names,
+        param_modules,
+        param_names,
+        buffer_module_names,
+        buffer_modules,
+        buffer_names
+    ):
         super(FunctionalModuleWithBuffers, self).__init__()
         self.stateless_model = stateless_model
+        self.param_module_names = param_module_names
+        self.param_modules = param_modules
         self.param_names = param_names
+        self.buffer_module_names = buffer_module_names
+        self.buffer_modules = buffer_modules
         self.buffer_names = buffer_names
 
-        self.all_names_map = dict(param_names_map)
-        self.all_names_map.update(buffer_names_map)
-
     @staticmethod
-    def _create_from(model, disable_autograd_tracking=False):
+    def _create_from(model):
         # TODO: We don't need to copy the model to create a stateless copy
         model_copy = copy.deepcopy(model)
-        params, param_names, param_names_map = extract_weights(model_copy)
-        buffers, buffer_names, buffer_names_map = extract_buffers(model_copy)
+        param_container = list(extract_weights(model_copy))
+        if len(param_container):
+            param_module_names, param_modules, param_names, params = zip(*param_container)
+        else:
+            param_module_names, param_modules, param_names, params = tuple(), tuple(), tuple(), tuple()
```

Comment:
Previously, we guaranteed that params is equivalent to what gets returned by original_model.parameters(). After this change, is that still true?

(Side note) To be honest, we've been thinking of changing the API so that params isn't returned as a flat list; instead we probably want to return some sort of dictionary or object so that one can easily figure out which `params` corresponds to which parameters on the original module. This is something that a couple of users have asked us for. If we returned a dictionary then it doesn't matter that params isn't the same as what gets returned by original_module.parameters()

Reply:
Oh man this would be so great! I could definitely use that feature.

To be honest, I was thinking about using [TensorDict](https://github.com/facebookresearch/rl/blob/main/tutorials/tensordict.ipynb) from torchrl to pass params to functorch stateless modules. We could nest the dicts (eg d["module"]["param"] to d["module.param"]), expand the params, change device or whatever, in batch and with little or no effort since all those ops are built-in tensordict methods. I think there's a good synergy that we could get from TensorDict functorch. At the moment, TensorDict isn't torchscriptable though, I don't know how much trouble it is for you.
@nairbv